src/Firestore/DocumentReference.php

```diff
@@ -0,0 +1,374 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Firestore;
+
+use Google\Cloud\Core\DebugInfoTrait;
+use Google\Cloud\Core\Iterator\ItemIterator;
+use Google\Cloud\Core\Iterator\PageIterator;
+use Google\Cloud\Core\Exception\NotFoundException;
+use Google\Cloud\Core\Exception\ConflictException;
+use Google\Cloud\Firestore\Connection\ConnectionInterface;
+
+/**
+ * Represents a reference to a Firestore document.
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\Firestore\FirestoreClient;
+ *
+ * $firestore = new FirestoreClient();
+ * $document = $firestore->document('users/john');
+ * ```
+ */
+class DocumentReference
+{
+    use SnapshotTrait;
+    use DebugInfoTrait;
+
+    /**
+     * @var ConnectionInterface
+     */
+    private $connection;
+
+    /**
+     * @var ValueMapper
+     */
+    private $valueMapper;
+
+    /**
+     * @var CollectionReference
+     */
+    private $parent;
+
+    /**
+     * @var string
+     */
+    private $name;
+
+    /**
+     * @param ConnectionInterface $connection A Connection to Cloud Firestore.
+     * @param ValueMapper $valueMapper A Firestore Value Mapper.
+     * @param CollectionReference $parent The collection in which this document is contained.
+     * @param string $name The fully-qualified document name.
+     */
+    public function __construct(
+        ConnectionInterface $connection,
+        ValueMapper $valueMapper,
+        CollectionReference $parent,
+        $name
+    ) {
+        $this->connection = $connection;
+        $this->valueMapper = $valueMapper;
+        $this->parent = $parent;
+        $this->name = $name;
+    }
+
+    /**
+     * Returns the parent collection.
+     *
+     * Example:
+     * ```
+     * $parent = $document->parent();
+     * ```
+     *
+     * @return CollectionReference
+     */
+    public function parent()
+    {
+        return $this->parent;
+    }
+
+    /**
+     * Get the document name.
+     *
+     * Example:
+     * ```
+     * $name = $document->name();
+     * ```
+     *
+     * @return string
+     */
+    public function name()
+    {
+        return $this->name;
+    }
+
+    /**
+     * Get the document identifier (i.e. the last path element).
+     *
+     * Example:
+     * ```
+     * $id = $document->id();
+     * ```
+     *
+     * @return string
+     */
+    public function id()
+    {
+        return $this->pathId($this->name);
+    }
+
+    /**
+     * Create a new document in Firestore. If the document already exists, this method will fail.
+     *
+     * Example:
+     * ```
+     * $document->create([
+     *     'name' => 'John',
+     *     'country' => 'USA'
+     * ]);
+     * ```
+     *
+     * @param array $fields An array containing field names paired with their value.
+     *        Accepts a nested array, or a simple array of field paths.
+     * @param array $options Configuration Options.
```

Comment:
All of our methods which trigger a service request accept this `$options` parameter. This allows users to configure request options such as retries (or in rest, just about anything in the request) on a per-call basis.

If there are options which are specific to the request, we document them, as in `DocumentReference::set()` below.

Reply:
Sounds good. I would just like to avoid accepting preconditions  as a first-party citizen here (if there is a way to access the underlying proto and add them manually, then we don't have to prevent it).